### PR TITLE
arm integration

### PR DESCRIPTION
**Pull request by takekoputa**
I didn't add any asserts or debugging flag. I would do that later.

### CODE REVIEW

File: src/cpu/flexcpu/inflight_inst.cc [hunk 1]
```diff
@@ -288,6 +372,15 @@ InflightInst::commitToTC()
         const RegId& dst_reg = instRef->destRegIdx(dst_idx);
         const GenericReg& result = getResult(dst_idx);
```
Comment (takekoputa) [inline src/cpu/flexcpu/inflight_inst.cc]: Oh my, this line is misplaced ...

File: src/cpu/flexcpu/inflight_inst.hh [hunk 2]
```diff
@@ -644,6 +647,14 @@ class InflightInst : public ExecContext,
 
     void pipeTrace();
 
+    // Since there are predicated instructions, the producer of a register
+    // might be predicated to false, and we should not forward anything
+    // from the instruction. This function determines the instruction holding
+    // the valid result from the register dependency chain. If no such an
+    // instruction exists, the backing context is holding the valid result.
+    std::pair<std::shared_ptr<InflightInst>, int>
+    getRegProducer(const StaticInst* si, int op_idx) const;
```
Comment (powerjg) [inline src/cpu/flexcpu/inflight_inst.hh]: Let's talk about this too... I think we may need to accelerate this with a change in the dependency map.

File: src/cpu/flexcpu/flexcpu_thread.cc [hunk 3]
```diff
@@ -1729,9 +1786,6 @@ FlexCPUThread::sendToMemory(shared_ptr<InflightInst> inst_ptr,
                 static_pointer_cast<ExecContext>(inst_ptr),
                 inst_ptr->traceData(), data.get(), callback);
         }
-
-        if (inst_ptr == inflightInsts.front())
```
[the remark this answers is not in the document]
Comment (takekoputa) [reply src/cpu/flexcpu/flexcpu_thread.cc]: I believe when a younger instruction is going to be committed, this instruction will be committed too.

[the remark this answers is not in the document]
Comment (takekoputa) [reply src/cpu/flexcpu/flexcpu_thread.cc]: Ohh I see, if a store is the last instruction of the program, it won't be committed.

File: src/cpu/flexcpu/flexcpu_thread.cc [hunk 4]
```diff
@@ -350,9 +352,17 @@ FlexCPUThread::bufferInstructionData(Addr vaddr, uint8_t* data)
...
+    // An instruction could be committed if it hasn't been committed and
+    // either:
+    //   - the instruction is in complete stage (all regs are updated)
+    //   - not in complete stage but is a store instruction that is not a
+    // store conditional (a normal store do not update regs,
+    // while store conditional does)
     return !inst_ref.isCommitted() &&
            (inst_ref.isComplete()
-            || (inst_ref.isMemorying() && inst_ref.staticInst()->isStore()));
+            || (inst_ref.isMemorying() &&
```
[the remark this answers is not in the document]
Comment (takekoputa) [reply src/cpu/flexcpu/flexcpu_thread.cc]: We can have something like `non_conditional_store = isMemorying() || (isStore() && !isStoreConditional())`, and we can return `!inst_ref.isCommitted() && (inst_reg.isComplete() || non_conditional_store)`.

File: src/cpu/flexcpu/inflight_inst.cc [hunk 5]
```diff
@@ -113,6 +114,64 @@ void InflightInst::pipeTrace()
...
+    const shared_ptr<InflightInst> producer = source.producer.lock();
+
+    assert(!this->isSquashed());
+    assert(!producer || (producer && !producer->isSquashed()));
+
+    if (producer && !producer->readPredicate()) {
+        shared_ptr<InflightInst> inst_walker = producer;
+        auto target_phys_reg =
+                        backingContext->flattenRegId(si->srcRegIdx(op_idx));
+        while (true) {
```
[the remark this answers is not in the document]
Comment (powerjg) [reply src/cpu/flexcpu/inflight_inst.cc]: Let's make this recursive.

**Comment by powerjg** [comment]
Good find!

**Pull request closed without merging**